### PR TITLE
ZKMathquant/testinfra basic - basic rounds of smoke and unit tests 

### LINKED ISSUE #4: Test coverage missing
Important followup to vet some of our basic shared agents and environments and create a rudimentary testing suite. Would be great to also include codecov for a visual demonstration of coverage.

### PR DESCRIPTION
**Pull request by ZKMathquant**
## What does this change?
The basic version with 38 tests -smoke and unit still giving 78% coverage
is now updated and made compatible to the new test files add.
Now have 82 tests with 79% coverage.

This only has core tests without

extended_agents
extended_environments
correctness tests(convergence,check for properties,adversarial)
and without
parametrized tests

Next version would be advanced testinfra but without param.
And the finally with param
from
 two other branches
https://github.com/SPAR-S26-IBRL-Stream/Project-Newcomb/tree/ZKMathquant/testinfra_adv with additional tests

and
https://github.com/SPAR-S26-IBRL-Stream/Project-Newcomb/tree/ZKMathquant/testinfra_adv_param with additonal tests but parametrized
..............................................................................
## Where in the codebase?

Added test sub-dir
Updated workflow and Project config (pyproject.toml, .gitignore, .github/)
added sample coverage report
ibrl/environments/Coin_tossing_game_envs file renamed as .py

## Related issue or discussion thread
Link to a GitHub issue, Slack/Discord thread or architecture doc section. If no issue exists, briefly explain why this change is needed. Write "N/A" only if truly not applicable.

### How did you verify correctness?
Everything is reviewed and tested 

# Before adding new test files test_infrabayesian_beliefs.py and test_knightian_uncertainty.py

| Name | Stmts | Miss | Cover | Missing |
|---|---|---|---|---|
| ibrl/__init__.py | 8 | 0 | 100% | |
| ibrl/agents/__init__.py | 9 | 0 | 100% | |
| ibrl/agents/base.py | 21 | 2 | 90% | 35, 63 |
| ibrl/agents/base_greedy.py | 42 | 14 | 67% | 26, 44-46, 60-64, 73-79 |
| ibrl/agents/bayesian.py | 17 | 1 | 94% | 32 |
| ibrl/agents/exp3.py | 26 | 1 | 96% | 47 |
| ibrl/agents/experimental1.py | 11 | 5 | 55% | 15-19 |
| ibrl/agents/experimental2.py | 41 | 31 | 24% | 28-33, 36-54, 57-70, 73-76, 79 |
| ibrl/agents/experimental3.py | 22 | 13 | 41% | 24-28, 32-37, 40, 43 |
| ibrl/agents/q_learning.py | 23 | 3 | 87% | 29-30, 42 |
| ibrl/environments/__init__.py | 10 | 0 | 100% | |
| ibrl/environments/asymmetric_damascus.py | 6 | 1 | 83% | 13 |
| ibrl/environments/bandit.py | 11 | 0 | 100% | |
| ibrl/environments/base.py | 22 | 2 | 91% | 61, 71 |
| ibrl/environments/base_newcomb_like.py | 18 | 0 | 100% | |
| ibrl/environments/coordination.py | 6 | 1 | 83% | 12 |
| ibrl/environments/damascus.py | 6 | 0 | 100% | |
| ibrl/environments/newcomb.py | 6 | 0 | 100% | |
| ibrl/environments/policy_dependent_bandit.py | 9 | 3 | 67% | 13, 16-17 |
| ibrl/environments/switching.py | 24 | 13 | 46% | 17-19, 23-30, 33, 36-40 |
| ibrl/simulators/__init__.py | 2 | 0 | 100% | |
| ibrl/simulators/simulator.py | 37 | 2 | 95% | 45, 63 |
| ibrl/utils/__init__.py | 3 | 0 | 100% | |
| ibrl/utils/construction.py | 35 | 0 | 100% | |
| ibrl/utils/debug.py | 4 | 1 | 75% | 9 |
| ibrl/utils/sampling.py | 6 | 0 | 100% | |
| TOTAL | 425 | 93 | 78% | |

------------------------------
Summary Results: 39 passed in 0.36s

# After adding new test files test_infrabayesian_beliefs.py and test_knightian_uncertainty.py

test_infra_basic test coverage

| Module | Statements | Missing | Coverage | Uncovered Lines |
|--------|-----------|---------|----------|-----------------|
| ibrl/__init__.py | 8 | 0 | 100% | — |
| ibrl/agents/__init__.py | 11 | 0 | 100% | — |
| ibrl/agents/base.py | 22 | 2 | 91% | 37, 65 |
| ibrl/agents/base_greedy.py | 42 | 11 | 74% | 26, 44-46, 60-64, 76-79 |
| ibrl/agents/bayesian.py | 18 | 1 | 94% | 33 |
| ibrl/agents/bernoulli_bayesian.py | 20 | 1 | 95% | 32 |
| ibrl/agents/exp3.py | 27 | 1 | 96% | 48 |
| ibrl/agents/experimental1.py | 11 | 5 | 55% | 15-19 |
| ibrl/agents/experimental2.py | 42 | 32 | 24% | 28-33, 36-54, 57-71, 74-77, 80 |
| ibrl/agents/experimental3.py | 22 | 13 | 41% | 24-28, 32-37, 40, 44 |
| ibrl/agents/infrabayesian.py | 36 | 5 | 86% | 61, 75-78 |
| ibrl/agents/q_learning.py | 24 | 3 | 88% | 30-31, 43 |
| ibrl/environments/__init__.py | 11 | 0 | 100% | — |
| ibrl/environments/asymmetric_damascus.py | 6 | 1 | 83% | 13 |
| ibrl/environments/bandit.py | 11 | 0 | 100% | — |
| ibrl/environments/base.py | 27 | 2 | 93% | 67, 77 |
| ibrl/environments/base_newcomb_like.py | 17 | 0 | 100% | — |
| ibrl/environments/bernoulli_bandit.py | 15 | 1 | 93% | 28 |
| ibrl/environments/coin_tossing_game_envs.py | 9 | 9 | 0% | 1-26 |
| ibrl/environments/coordination.py | 6 | 1 | 83% | 12 |
| ibrl/environments/damascus.py | 6 | 0 | 100% | — |
| ibrl/environments/newcomb.py | 6 | 0 | 100% | — |
| ibrl/environments/policy_dependent_bandit.py | 9 | 3 | 67% | 13, 16-17 |
| ibrl/environments/switching.py | 24 | 13 | 46% | 17-19, 23-30, 33, 36-40 |
| ibrl/infrabayesian/__init__.py | 4 | 0 | 100% | — |
| ibrl/infrabayesian/a_measure.py | 15 | 2 | 87% | 21, 28 |
| ibrl/infrabayesian/beliefs.py | 85 | 19 | 78% | 22, 32, 43, 48, 66, 88, 101-103, 106-110, 113, 124, 130-134 |
| ibrl/infrabayesian/infradistribution.py | 41 | 3 | 93% | 52, 61, 93 |
| ibrl/outcome.py | 5 | 0 | 100% | — |
| ibrl/simulators/__init__.py | 2 | 0 | 100% | — |
| ibrl/simulators/simulator.py | 37 | 2 | 95% | 45, 63 |
| ibrl/utils/__init__.py | 3 | 0 | 100% | — |
| ibrl/utils/construction.py | 49 | 10 | 80% | 35-36, 87-95 |
| ibrl/utils/debug.py | 4 | 1 | 75% | 9 |
| ibrl/utils/sampling.py | 6 | 0 | 100% | — |
| **TOTAL** | **681** | **141** | **79%** | — |

**Test Results:** 82 passed, 2 warnings

### What is the math or theory behind this change?
There is math behind the additional two tests indeed,especially the blowup warnings.It is covered in the respective docs.

## If experiment (`experiments/`):

Does your experiment directory have a `README.md`? Your README should cover what the experiment is, why you ran it, results and your interpretation.
- The README is updated with the commands on how to run the tests locally

### CODE REVIEW

File: tests/test_simulator.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        assert results["rewards"].shape == (num_runs, num_steps)
+
+    def test_simulate_rewards_valid(self, num_actions, seed):
+        env = BanditEnvironment(num_actions=num_actions, seed=seed)
+        agent = QLearningAgent(num_actions=num_actions, seed=seed + 1)
+        options = {"num_steps": 10, "num_runs": 2}
+        results = simulate(env, agent, options)
+        
+        assert np.all(np.isfinite(results["rewards"]))
+        assert np.all(np.isfinite(results["average_reward"]))
```
Comment (alatimer) [inline tests/test_simulator.py]: Probably all three of these tests could/should be a single test with all the assertions, since we are just running the same simulation and checking different aspects of it.  (don't need to run it three times independently)

**Review by alatimer — APPROVED** [review]
Looks great! I've added some comments you might want to consider to prune the tests that feel redundant. Thanks for the contribution.

**Comment by redradman** [comment]
@claude do a code review here

**Comment by ZKMathquant** [comment]
The PR is updated with @alatimer and @claude 's PR review feedbacks.
Feel free to check.

@claude once aagin do a code review here

**Comment by ZKMathquant** [comment]
Taken note of and made changes except

- 10^-9 or -3 is a researcher choice made by @fllor or @alatimer ,leaving it to them
- the unused imports were not unused though @ claude bot folk,I removed them and 9-10 tests failed.

Changed everything else

@claude do PR review once again,should I merge?

File: tests/test_agents.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    def test_update(self, q_learning_agent):
+        probs = q_learning_agent.get_probabilities()
+        action = 0
+        reward = 1.0
+        outcome = Outcome(reward=reward, env_action=None)
+        q_learning_agent.update(probs, action, outcome)
+        assert q_learning_agent.step == 2
+        assert q_learning_agent.q[action] > 0
+
+    def test_learning_rate_none(self, num_actions, seed):
```
[the remark this answers is not in the document]
Comment (alatimer) [reply tests/test_agents.py]: I see more substance was added, but now name should be changed to reflect what is actually being tested

**Review by alatimer — APPROVED** [review]
Left one comment, other than that LGTM

File: tests/test_agents.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZKMathquant) [reply tests/test_agents.py]: it is commented out right
```
#It verifies that when learning_rate=None, the Q-learning agent uses sample averaging
# and correctly increments the action count after an update
```
The name is kept unchanges as it reflects exactly that no.
If you follow through the flow of claude suggestions,you would find the idea behind changes made.

File: tests/test_environments.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        assert isinstance(optimal, (int, float, np.integer, np.floating))
+        assert optimal >= 0
+
+
+class TestDeathInDamascusEnvironment:
+    def test_initialization(self, seed):
+        env = DeathInDamascusEnvironment(num_actions=2, seed=seed)
+        assert env.num_actions == 2
+
+    def test_reward_table_structure(self, damascus_env):
```
[the remark this answers is not in the document]
Comment (ZKMathquant) [reply tests/test_environments.py]: They were removed in the recent.

File: tests/test_simulator.py (hunk 1 above)
Comment (ZKMathquant) [reply tests/test_simulator.py:28]: It is taken care of in the recent

File: tests/test_environments.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    def test_initialization(self, seed):
+        env = NewcombEnvironment(num_actions=2, seed=seed)
+        assert env.num_actions == 2
+        assert env.reward_table.shape == (2, 2)
+
+    def test_reward_table_structure(self, newcomb_env):
+        assert newcomb_env.reward_table[0, 0] == 10  # boxB
+        assert newcomb_env.reward_table[0, 1] == 15  # boxB + boxA
+        assert newcomb_env.reward_table[1, 0] == 0
+        assert newcomb_env.reward_table[1, 1] == 5   # boxA
```
[the remark this answers is not in the document]
Comment (ZKMathquant) [reply tests/test_environments.py]: Hard coding is avoided in the recent

**Comment by ZKMathquant** [comment]
@claude can you describe the changes made in the recent 3 commits into this PR to @alatimer for briefly getting used to deletions/update/additions during that along  what part of the PR review feedbacks they addressed 

Also

it shows

```
skipped checks
[Claude Code / claude (pull_request_review)](https://github.com/SPAR-S26-IBRL-Stream/Project-Newcomb/actions/runs/24209143842/job/70672870513?pr=25)
Claude Code / claude (pull_request_review)Skipped 1 minute ago
successful checks
[Tests / Smoke Tests (Round 1) (pull_request)](https://github.com/SPAR-S26-IBRL-Stream/Project-Newcomb/actions/runs/24193908907/job/70618388586?pr=25)
Tests / Smoke Tests (Round 1) (pull_request)Successful in 11s
[Tests / Unit Tests (Round 2) (pull_request)](https://github.com/SPAR-S26-IBRL-Stream/Project-Newcomb/actions/runs/24193908907/job/70618388580?pr=25)
Tests / Unit Tests (Round 2) (pull_request)Successful in 12s
```
why does it skip?is that a bug or a feature?

File: tests/test_agents.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alatimer) [reply tests/test_agents.py]: This is an unnecessary assertion: assert agent.learning_rate is None
Since we are explicitly setting that in agent = QLearningAgent(num_actions=num_actions, learning_rate=None, seed=seed)

Really, what this test is now doing, is what is commented out, that the Q learning agent correctly increments the action count after an update:
        probs = agent.get_probabilities()
        outcome = Outcome(reward=1.0, env_action=None)
        agent.update(probs, 0, outcome)
        assert agent.counts[0] == 1

So the new test should be named eg "test_qlearner_correctly_increments_action_count"

If it is testing that it uses sample averaging, I am not seeing that in the code.

My main general point is that test names should be descriptive of what exactly is being tested. It makes debugging and readability/maintainability easier.

**Comment by alatimer** [comment]
@ZKMathquant I had already seen the 3 most recent changes and was saying that they look good. This was my only outstanding comment: https://github.com/SPAR-S26-IBRL-Stream/Project-Newcomb/pull/25#discussion_r3060113371
I have added a reply to that thread clarifying my point, which stands with the most recent changes.

**Comment by ZKMathquant** [comment]
> @ZKMathquant I had already seen the 3 most recent changes and was saying that they look good. This was my only outstanding comment: [#25 (comment)](https://github.com/SPAR-S26-IBRL-Stream/Project-Newcomb/pull/25#discussion_r3060113371) I have added a reply to that thread clarifying my point, which stands with the most recent changes.

Got it.
@claude has this suggestion been taken care of in the recent commit?

**Comment by ZKMathquant** [comment]
> **Claude finished @ZKMathquant's task in 53s** —— [View job](https://github.com/SPAR-S26-IBRL-Stream/Project-Newcomb/actions/runs/24231610225)
> 
> * [x]  Read `tests/test_agents.py` to check current state
> * [x]  Compare against @alatimer's suggestion
> * [x]  Report findings
> 
> ### Status of @alatimer's Suggestion
> @alatimer had two points in their outstanding comment:
> 
> Point	Status
> Remove `assert agent.learning_rate is None` (unnecessary since it's explicitly set)	✅ **Done** — that assertion is gone from the current file
> Rename test to something descriptive like `test_qlearner_correctly_increments_action_count`	✅ **Partially done** — renamed to `test_sample_average_increments_action_count` (commit `561083a`)
> The `assert agent.learning_rate is None` line no longer exists in the file. The test at line 35 now reads:
> 
> ```python
> def test_sample_average_increments_action_count(self, num_actions, seed):
>     """Verify Q-learning agent increments action count in sample average mode."""
>     agent = QLearningAgent(num_actions=num_actions, learning_rate=None, seed=seed)
>     agent.reset()
>     assert hasattr(agent, 'counts')
>     
>     probs = agent.get_probabilities()
>     outcome = Outcome(reward=1.0, env_action=None)
>     agent.update(probs, 0, outcome)
>     assert agent.counts[0] == 1
> ```
> 
> **The only difference from @alatimer's exact suggestion** is the name: `test_sample_average_increments_action_count` vs their suggested `test_qlearner_correctly_increments_action_count`. Both are descriptive — the chosen name emphasizes the _mode_ (sample averaging) while @alatimer's suggestion emphasizes the _agent_ (Q-learner). Either is acceptable; it's a minor stylistic preference.
> 
> If @alatimer is happy with `test_sample_average_increments_action_count` as sufficiently descriptive, this is resolved and the PR is ready to merge.

@alatimer I am going ahead and merging in that case.As the name can be changed in any of the next PRs anyway.
